[Navigation] Implement navigation.updateCurrentEntry()

TingPing opened the pull request:
#### 6207ac0f032490e8fb53a51e166684b3f2edccc0
<pre>
[Navigation] Implement navigation.updateCurrentEntry()
<a href="https://bugs.webkit.org/show_bug.cgi?id=268261">https://bugs.webkit.org/show_bug.cgi?id=268261</a>

Reviewed by Alex Christensen.

This implements `navigation.UpdateCurrentEntry()`, both the method storing state, and emitting the currententryupdated event.

<a href="https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-navigation-updatecurrententry">https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-navigation-updatecurrententry</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/anchor-click-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/constructor-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/history-back-same-doc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/history-pushState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/history-replaceState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/location-api-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigation-back-forward-same-doc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigation-navigate-preventDefault-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigation-navigate-replace-same-doc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigation-navigate-same-doc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigation-updateCurrentEntry-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/cross-document-away-and-back-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/cross-document-location-api-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/exception-order-initial-about-blank-unserializablestate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/exception-order-not-fully-active-unserializablestate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/history-pushState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/history-replaceState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/location-reload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/not-fully-active-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/opaque-origin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/same-document-away-and-back-location-api-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/unserializable-expected.txt:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::updateCurrentEntry):
* Source/WebCore/page/Navigation.h:
* Source/WebCore/page/Navigation.idl:
* Source/WebCore/page/NavigationHistoryEntry.cpp:
(WebCore::NavigationHistoryEntry::getState const):
* Source/WebCore/page/NavigationHistoryEntry.h:
* Source/WebCore/page/NavigationHistoryEntry.idl:

Canonical link: <a href="https://commits.webkit.org/274321@main">https://commits.webkit.org/274321@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c409fda8b4fd0d10853f44dc1b8779b6c807c2c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38607 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17539 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40936 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41141 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/34270 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/40912 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20340 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14883 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32462 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39180 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14787 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33590 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12846 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12834 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34450 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42417 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35078 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/34844 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38678 "Passed tests") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13431 "Build is being retried. Recent messages:OS: Sonoma (14.0), Xcode: 15.0; Skipping applying patch since patch_id isn't provided; Checked out pull request; compiling") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11114 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36872 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15040 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/13891 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5047 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14515 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rwlbuis reviewed (commented):
Can you add a spec link in the commit message and before the method?

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -130,8 +134,36 @@ Navigation::Result Navigation::forward(Options&&, Ref<DeferredPromise>&& committ
+    // Not explicitly in the spec, but expected by web-platform-tests/navigation-api/updateCurrentEntry-method/exception-order-initial-about-blank-unserializablestate.html
+    bool isAboutBlank = window()->frame()->document()->url().isAboutBlank();
+
+    auto serializedState = SerializedScriptValue::create(globalObject, options.state, SerializationForStorage::Yes, isAboutBlank ? SerializationErrorMode::NonThrowing : SerializationErrorMode::Throwing);
```
Can't we just always use the variant that returns ExceptionOr, and return that if there is an exception? Chromium code does not seem to have special code for about:blank at this level.

TingPing replied:
> Chromium code does not seem to have special code for about:blank at this level.

I guess their serializer just returns a different exception when ran in the context of a blank page and they codified that into the WPT test. It explicitly tests that it returns InvalidState instead of DataCloneError.

TingPing replied:
> Can't we just always use the variant that returns ExceptionOr, and return that if there is an exception?

This is a void function, so I don't think it looks any nicer anyway.

```cpp
    Vector<RefPtr<MessagePort>> dummyPorts;
    auto serializeResult = SerializedScriptValue::create(globalObject, options.state, { }, dummyPorts, SerializationForStorage::Yes);
    if (serializeResult.hasException()) {
        // Not explicitly in the spec, but expected by web-platform-tests/navigation-api/updateCurrentEntry-method/exception-order-initial-about-blank-unserializablestate.html
        if (window()->frame()->document()->url().isAboutBlank())
            return Exception { ExceptionCode::InvalidStateError };
        return serializeResult.releaseException();
    }
```

rwlbuis replied:
Do you know where in the specification the special behaviour of about:blank is described? I am not convinced yet it should be part of this change.

TingPing replied:
I did not find it in the spec, only in WPT. https://github.com/web-platform-tests/wpt/blob/master/navigation-api/updateCurrentEntry-method/exception-order-initial-about-blank-unserializablestate.html

And there is no descriptive history there of why.

rwlbuis replied:
Thanks, I think it is best to contact some people about why that test is that way, and until we understand and agree with the need of it, to leave the about:blank part out.

TingPing replied:
I've removed it from this patch for now.

rwlbuis commented on Source/WebCore/page/NavigationHistoryEntry.h:
```diff
@@ -64,7 +69,7 @@ class NavigationHistoryEntry final : public RefCounted<NavigationHistoryEntry>,
     const WTF::UUID m_id;
     uint64_t m_index;
-    JSC::JSValue m_state;
+    RefPtr<SerializedScriptValue> m_serializedState;
```
Have you condered storing the state on the HistoryItem? It seems to be what the spec and chromium implementation do.

TingPing replied:
Done.

rwlbuis commented on Source/WebCore/history/HistoryItem.cpp:
```diff
@@ -306,6 +307,11 @@ void HistoryItem::setStateObject(RefPtr<SerializedScriptValue>&& object)
     notifyChanged();
 }
 
+void HistoryItem::setNavigationAPIStateObject(RefPtr<SerializedScriptValue>&& object)
```
Can also use a spec link :)

rwlbuis commented on Source/WebCore/history/HistoryItem.h:
```diff
@@ -135,6 +135,9 @@ class HistoryItem : public RefCounted<HistoryItem>, public CanMakeWeakPtr<Histor
     WEBCORE_EXPORT void setStateObject(RefPtr<SerializedScriptValue>&&);
     SerializedScriptValue* stateObject() const { return m_stateObject.get(); }
 
+    WEBCORE_EXPORT void setNavigationAPIStateObject(RefPtr<SerializedScriptValue>&&);
```
WEBCORE_EXPORT is for WebKit/ land IIRC, it is likely not needed right now.

rwlbuis commented on Source/WebCore/page/NavigationHistoryEntry.h:
```diff
@@ -28,6 +28,8 @@
 #include "EventHandler.h"
 #include "EventTarget.h"
+#include "HistoryItem.h"
+#include "SerializedScriptValue.h"
```
This can likely just be a forward reference. In general I'd say check all includes, particularly in the header where forward references are preferred.

rwlbuis commented on Source/WebCore/page/NavigationHistoryEntry.h:
```diff
@@ -64,7 +72,8 @@ class NavigationHistoryEntry final : public RefCounted<NavigationHistoryEntry>,
     uint64_t m_index;
-    JSC::JSValue m_state;
+    // TODO: Every entry is supposed to have an associated history item.
+    std::optional<Ref<HistoryItem>> m_associatedHistoryItem;
```
I wonder if we'll need a WeakRef instead. Maybe we can wait until we add a PR that handles reactions to HistoryItem changes (deletion etc) though. Does the spec have guidelines for that?

TingPing replied:
I read it as both are owned by the "Navigable" so would have the same lifetimes.

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -130,8 +134,30 @@ Navigation::Result Navigation::forward(Options&&, Ref<DeferredPromise>&& committ
+    if (!current)
+        return Exception { ExceptionCode::InvalidStateError };
+
+    auto serializedState = SerializedScriptValue::create(globalObject, options.state, SerializationForStorage::Yes, SerializationErrorMode::Throwing);
```
I think we may want to use Nonthrowing for now or do something testing if there was an exceptions. The way the code is now seems like we ignore the exception. Some FIXME is also an option.

TingPing replied:
WPT has tests covering this and it works as expected.

> PASS updateCurrentEntry() must throw if state is unserializable (WritableStream)

rwlbuis reviewed (commented):
Patch looks quite good to me now, I'd like to hear from @achristensen07 too.

TingPing committed:
[Navigation] Implement navigation.updateCurrentEntry()
https://bugs.webkit.org/show_bug.cgi?id=268261

Reviewed by Alex Christensen.

This implements `navigation.UpdateCurrentEntry()`, both the method storing state, and emitting the currententryupdated event.

https://html.spec.whatwg.org/multipage/nav-history-apis.html#dom-navigation-updatecurrententry

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/anchor-click-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/constructor-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/history-back-same-doc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/history-pushState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/history-replaceState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/location-api-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigation-back-forward-same-doc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigation-navigate-preventDefault-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigation-navigate-replace-same-doc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigation-navigate-same-doc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigation-updateCurrentEntry-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/cross-document-away-and-back-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/cross-document-location-api-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/exception-order-initial-about-blank-unserializablestate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/exception-order-not-fully-active-unserializablestate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/history-pushState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/history-replaceState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/location-reload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/not-fully-active-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/opaque-origin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/same-document-away-and-back-location-api-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/unserializable-expected.txt:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::updateCurrentEntry):
* Source/WebCore/page/Navigation.h:
* Source/WebCore/page/Navigation.idl:
* Source/WebCore/page/NavigationHistoryEntry.cpp:
(WebCore::NavigationHistoryEntry::getState const):
* Source/WebCore/page/NavigationHistoryEntry.h:
* Source/WebCore/page/NavigationHistoryEntry.idl:

Canonical link: https://commits.webkit.org/274321@main
